app-engine/java/src/com/google/cloud/tools/intellij/appengine/java/facet/standard/impl/AppEngineStandardCommunityWebIntegration.java

```diff
@@ -49,10 +52,27 @@
   @Override
   public VirtualFile suggestParentDirectoryForAppEngineWebXml(
       @NotNull Module module, @NotNull ModifiableRootModel rootModel) {
-    final VirtualFile root = ArrayUtil.getFirstElement(rootModel.getContentRoots());
+    BuildSystemAppEngineWebXmlDirectoryProvider[] appEngineWebXmlDirectoryProviders =
+        Extensions.getExtensions(BuildSystemAppEngineWebXmlDirectoryProvider.EP_NAME);
+    for (BuildSystemAppEngineWebXmlDirectoryProvider provider : appEngineWebXmlDirectoryProviders) {
+      Optional<String> pathOptional = provider.getAppEngineWebXmlDirectoryPath(module);
```

Comment:
It seems a bit odd to iterate through these rather than pick the right provider for the kind of module this is.

Reply:
each provider will either return Optional#empty if the given module is not applicable to the given build system.

i.e. if the module is not a Maven module, then `MavenAppEngineWebXmlDirectoryProvider` will return `Optional#empty`.

I agree it feels a bit odd to iterate through them all, but the benefit of this is that it allows us to have a single extension point and have each extension determine if its applicable, rather than the other way around.

I think a way to improve readablility would be to add an

`isApplicable(Module module)`

method to the interface. And instead of iterating blindly through each provider, first check `isApplicable`. I'll make this update

Reply:
on second though. I don't like extracting an `isApplicable` method since it leaves open the possibility of calling the interface method `getAppEngineWebXmlDirectoryPath` in an invalid state. I think its better to have the `getAppEngineWebXmlDirectoryPath` confirm itself that its applicable for the build system and return empty if not (as it is now). wdyt?

Reply:
In this case it feels we can have multiple build systems support for a module, and the one selected is completely random. It slightly feels we miss module type safety checking or something similar. Does a module have something that uniquely identifies its build system for this case, could we map this to build system provider implementation?